edb/lib/net.edgeql

```diff
@@ -0,0 +1,79 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2024-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+CREATE MODULE net;
+
+CREATE SCALAR TYPE net::RequestState EXTENDING std::enum<
+    Pending,
+    InProgress,
+    Completed,
+    Failed
+>;
+
+CREATE SCALAR TYPE net::RequestFailureKind EXTENDING std::enum<
+    NetworkError,
+    Timeout
```

Comment:
My concern here is that I - as a user - don't know if I should re-schedule my request upon these errors, like `NetworkError` could be `cannot resolve hostname` (e.g. should backoff for some time and retry) or `connection reset receiving response` (e.g. should check remote first to avoid double transaction or something), `Timeout` may also happen before or after sending a request.

Regex-matching the `message` is suboptimal - e.g. `LANG=zh_CN.UTF-8` may localize the message in some cases of propagated low-level errors. I know we decided to let user decide on retrying, but I currently don't have a better suggestion to expose more info about the error (so that the user could actually decide) while keeping the API simple. So I guess it's fine for now.

Reply:
Oh, we should absolutely have a lot more enums here, but I'm not even sure what all of the different error possibilities are yet. Mostly I imagine these are the rough groups of kinds of errors we could get from the underlying library: either they've timed out (which we would control I think?) or the request library (httpx in python or reqwest in rust) would throw an exception and we'd bubble that back up. Makes sense to match the granularity of the underlying library to make it easiest for the consumer.